File: core/trips.js
/* Trip / drive-segment tracking — shared by the MagicMirror node_helper and the
 * HA coordinator, both of which drive it with their own streaming state + storage.
 *
 * A "trip" runs from a parked anchor until the car is parked again. Distance is
 * the odometer delta (works even in cached-poll mode, where per-drive engine
 * state is never seen); energy is the state-of-charge drop across the trip,
 * converted with the usable pack size. If the car charged during the window the
 * energy / efficiency / cost fields are left null — the distance still counts.
 *
 * Pure: no DOM, no `this`, no clock of its own (the caller passes `t`).
 */
(function (root, factory) {
  if (typeof module === "object" && module.exports) module.exports = factory();
  else root.KiaAccessTrips = factory();
})(typeof self !== "undefined" ? self : this, function () {
  "use strict";

  var DEFAULT_CAPACITY_KWH = 99.8; // Kia EV9 usable
  var MIN_KM = 0.5;                // ignore driveway shuffles below this
  var PARK_GAP_MIN = 8;            // odo stable / engine off this long -> parked
  var MI_PER_KM = 0.621371;

  function num(v) {
    if (v == null || v === "") return null;
    var n = Number(v);
    return isFinite(n) ? n : null;
  }
  function round(n, dp) {
    if (n == null) return null;
    var f = Math.pow(10, dp == null ? 2 : dp);
    return Math.round(n * f) / f;
  }
  function haversineKm(aLat, aLon, bLat, bLon) {
    if (aLat == null || aLon == null || bLat == null || bLon == null) return null;
    // null already handled; also reject NaN/Infinity (both pass a bare
    // != null check and their trig math can produce NaN/Infinity results)
    // and an out-of-range-but-finite lat/lon -- the one caller here
    // (close()) already passes a null result through round() safely, so
    // extending this is a no-op for that call site and closes the gap for
    // anything else that might reach this function.
    if (typeof aLat !== "number" || typeof aLon !== "number" ||
        typeof bLat !== "number" || typeof bLon !== "number") return null;
    if (!isFinite(aLat) || !isFinite(aLon) || !isFinite(bLat) || !isFinite(bLon)) return null;
    if (aLat < -90 || aLat > 90 || bLat < -90 || bLat > 90) return null;
    if (aLon < -180 || aLon > 180 || bLon < -180 || bLon > 180) return null;
    var R = 6371, p = Math.PI / 180;
    var dLat = (bLat - aLat) * p, dLon = (bLon - aLon) * p;
    var s = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
      Math.cos(aLat * p) * Math.cos(bLat * p) * Math.sin(dLon / 2) * Math.sin(dLon / 2);
    return R * 2 * Math.asin(Math.sqrt(s));
  }

  function close(open, opts, endAt) {
    var cap = num(opts.capacityKwh) || DEFAULT_CAPACITY_KWH;
    var price = num(opts.pricePerKwh) || 0;
    var dist = (open.lastOdo != null && open.anchorOdo != null)
      ? Math.max(0, open.lastOdo - open.anchorOdo) : null;
    if (dist == null || dist < (num(opts.minKm) != null ? num(opts.minKm) : MIN_KM)) {
      return null;
    }
    var usedPct = (!open.chargedSince && open.anchorPct != null && open.lastPct != null)
      ? open.anchorPct - open.lastPct : null;
    var kwh = (usedPct != null && usedPct > 0) ? (usedPct / 100) * cap : null;
    var mins = Math.max(1, Math.round((endAt - open.anchorAt) / 60000));
    return {
      startedAt: open.anchorAt,
      endedAt: endAt,
      minutes: mins,
      distanceKm: round(dist, 1),
      distanceMi: round(dist * MI_PER_KM, 1),
      usedPct: usedPct != null ? round(usedPct, 1) : null,
      // raw start/end % (not just the usedPct delta) -- core/analytics.js's
      // rangeAccuracy() needs the actual starting % to compare against
      // what Kia displayed AT that %, not just how much was consumed.
      startPct: open.anchorPct != null ? round(open.anchorPct, 1) : null,
      endPct: open.lastPct != null ? round(open.lastPct, 1) : null,
      kwh: round(kwh, 2),
      // efficiency both ways round — mi/kWh reads high-is-good, kWh/100mi low-is-good
      miPerKwh: (kwh != null && kwh > 0) ? round(dist * MI_PER_KM / kwh, 2) : null,
      kwhPer100mi: (kwh != null && dist > 0)
        ? round(kwh / (dist * MI_PER_KM) * 100, 1) : null,
      cost: (kwh != null && price > 0) ? round(kwh * price, 2) : null,
      pricePerKwh: price || null,
      fromLat: open.anchorLat, fromLon: open.anchorLon,
      toLat: open.lastLat, toLon: open.lastLon,
      straightLineKm: round(
        haversineKm(open.anchorLat, open.anchorLon, open.lastLat, open.lastLon), 1),
      chargedDuring: !!open.chargedSince,
      // Kia's own displayed EV range at the trip's START (the anchor) --
      // core/analytics.js's rangeAccuracy() compares this against what
      // this trip's own observed efficiency implies was actually
      // achievable from that same starting %. null if the car didn't
      // report a range reading at that moment (e.g. combustion/fuel-only
      // range showing instead, or a momentary gap in the poll).
      startRangeKm: open.anchorRangeKm != null ? round(open.anchorRangeKm, 1) : null,
      // outside temperature at the trip's start -- core/analytics.js
      // buckets observed efficiency by this. Average of start+end would
      // blur a trip that started cold and warmed up; the START reading is
      // what the driver actually experienced deciding whether to trust
      // the car's range estimate that morning.
      outsideTempC: open.anchorTempC != null ? round(open.anchorTempC, 1) : null
    };
  }

  /**
   * Fold one state sample into trip tracking.
   * @param {object|null} open  trip in progress, or null
   * @param {object} cur  { t, odometerKm, batteryPct, charging, carOn,
   *                         locationLat, locationLon, rangeKm, outsideTempC }
   *   rangeKm/outsideTempC are optional -- only used by core/analytics.js's
   *   rangeAccuracy()/observedEfficiency() temperature buckets; everything
   *   else here works exactly as before when they're omitted.
   * @param {object} opts { pricePerKwh, capacityKwh, minKm, parkGapMin }
   * @returns {{ open: (object|null), closed: (object|null) }}
   */
  function update(open, cur, opts) {
    opts = opts || {};
    var t = num(cur.t) || Date.now();
    var odo = num(cur.odometerKm);
    var pct = num(cur.batteryPct);
    var lat = num(cur.locationLat);
    var lon = num(cur.locationLon);
    var rangeKm = num(cur.rangeKm);
    var tempC = num(cur.outsideTempC);
    var charging = cur.charging === true;
    var carOn = cur.carOn === true;
    var gapMs = (num(opts.parkGapMin) || PARK_GAP_MIN) * 60000;

    if (odo == null) return { open: open, closed: null };

    // no trip yet: set / refresh the parked anchor
    if (!open) {
      return {
        open: {
          anchorOdo: odo, anchorPct: pct, anchorAt: t,
          anchorLat: lat, anchorLon: lon,
          anchorRangeKm: rangeKm, anchorTempC: tempC,
          lastOdo: odo, lastPct: pct, lastAt: t,
          lastLat: lat, lastLon: lon,
          lastRangeKm: rangeKm, lastTempC: tempC,
          movedAt: t, chargedSince: false
        },
        closed: null
      };
    }

    var moved = odo > open.lastOdo + 0.05;
    if (charging || (pct != null && open.lastPct != null && pct > open.lastPct + 1)) {
      open.chargedSince = true;
    }

    if (moved) {
      open.lastOdo = odo;
      open.movedAt = t;
      if (pct != null) open.lastPct = pct;
      if (lat != null) { open.lastLat = lat; open.lastLon = lon; }
      if (rangeKm != null) open.lastRangeKm = rangeKm;
      if (tempC != null) open.lastTempC = tempC;
      open.lastAt = t;
      return { open: open, closed: null };
    }

    // not moving this sample
    if (pct != null) open.lastPct = pct;
    if (rangeKm != null) open.lastRangeKm = rangeKm;
    if (tempC != null) open.lastTempC = tempC;
    if (lat != null && open.lastOdo === open.anchorOdo) {
      // still parked at the anchor — keep the anchor fresh so a later charge
      // there is attributed correctly
      open.anchorLat = lat; open.anchorLon = lon;
      if (pct != null) open.anchorPct = pct;
      if (rangeKm != null) open.anchorRangeKm = rangeKm;
      if (tempC != null) open.anchorTempC = tempC;
    }
    open.lastAt = t;

    var parked = (!carOn && t - open.movedAt >= gapMs);
    if (parked && open.lastOdo > open.anchorOdo) {
      var trip = close(open, opts, open.movedAt);
      return {
        open: {
          anchorOdo: open.lastOdo, anchorPct: open.lastPct, anchorAt: t,
          anchorLat: open.lastLat, anchorLon: open.lastLon,
          anchorRangeKm: open.lastRangeKm, anchorTempC: open.lastTempC,
          lastOdo: open.lastOdo, lastPct: open.lastPct, lastAt: t,
          lastLat: open.lastLat, lastLon: open.lastLon,
          lastRangeKm: open.lastRangeKm, lastTempC: open.lastTempC,
          movedAt: t, chargedSince: false
        },
        closed: trip
      };
    }
    if (parked) {
      // parked but no distance (or below MIN_KM) — reset the anchor here
      open.anchorOdo = open.lastOdo;
      open.anchorPct = open.lastPct;
      open.anchorAt = t;
      open.anchorLat = open.lastLat;
      open.anchorLon = open.lastLon;
      open.anchorRangeKm = open.lastRangeKm;
      open.anchorTempC = open.lastTempC;
      open.chargedSince = false;
    }
    return { open: open, closed: null };
  }

  /** totals over the last `days` (default 30) of a trip list */
  function summary(trips, days) {
    var cutoff = Date.now() - (days || 30) * 864e5;
    var km = 0, kwh = 0, cost = 0, n = 0, haveCost = false, haveKwh = false;
    (trips || []).forEach(function (tr) {
      if (!tr || num(tr.endedAt) == null || tr.endedAt < cutoff) return;
      n += 1;
      if (tr.distanceKm != null) km += tr.distanceKm;
      if (tr.kwh != null) { kwh += tr.kwh; haveKwh = true; }
      if (tr.cost != null) { cost += tr.cost; haveCost = true; }
    });
    var mi = km * MI_PER_KM;
    return {
      count: n,
      distanceKm: round(km, 1),
      distanceMi: round(mi, 1),
      kwh: haveKwh ? round(kwh, 1) : null,
      cost: haveCost ? round(cost, 2) : null,
      miPerKwh: (haveKwh && kwh > 0) ? round(mi / kwh, 2) : null,
      costPerMi: (haveCost && mi > 0) ? round(cost / mi, 3) : null
    };
  }

  return {
    update: update,
    summary: summary,
    haversineKm: haversineKm,
    DEFAULT_CAPACITY_KWH: DEFAULT_CAPACITY_KWH,
    MI_PER_KM: MI_PER_KM
  };
});

File: trips.py
"""Trip / drive-segment tracking - Python port of core/trips.js.

Kept line-for-line with the JS; both are exercised in CI
(test/trips.test.js / test/trips_test.py).
"""
from __future__ import annotations

import math
import time

DEFAULT_CAPACITY_KWH = 99.8  # Kia EV9 usable
MIN_KM = 0.5
PARK_GAP_MIN = 8
MI_PER_KM = 0.621371


def _num(v):
    if v is None or v == "":
        return None
    try:
        f = float(v)
    except (TypeError, ValueError):
        return None
    return f if f == f and f not in (float("inf"), float("-inf")) else None


def _round(n, dp=2):
    if n is None:
        return None
    f = 10 ** dp
    return round(n * f) / f


def haversine_km(a_lat, a_lon, b_lat, b_lon):
    # None already handled; also reject NaN/Infinity (both are real floats
    # that pass a bare None check, and their trig math can raise a "math
    # domain error" downstream) and an out-of-range-but-finite lat/lon --
    # the one caller here (_close()) already passes a None result through
    # _round() safely, so extending this is a no-op for that call site and
    # closes the gap for anything else that might reach this function.
    for v in (a_lat, a_lon, b_lat, b_lon):
        if not isinstance(v, (int, float)) or isinstance(v, bool) or not math.isfinite(v):
            return None
    if not (-90 <= a_lat <= 90 and -90 <= b_lat <= 90):
        return None
    if not (-180 <= a_lon <= 180 and -180 <= b_lon <= 180):
        return None
    r, p = 6371, math.pi / 180
    d_lat = (b_lat - a_lat) * p
    d_lon = (b_lon - a_lon) * p
    s = (math.sin(d_lat / 2) ** 2
         + math.cos(a_lat * p) * math.cos(b_lat * p) * math.sin(d_lon / 2) ** 2)
    return r * 2 * math.asin(math.sqrt(s))


def _close(open_t, opts, end_at):
    cap = _num(opts.get("capacityKwh")) or DEFAULT_CAPACITY_KWH
    price = _num(opts.get("pricePerKwh")) or 0
    dist = (max(0, open_t["lastOdo"] - open_t["anchorOdo"])
            if (open_t.get("lastOdo") is not None and open_t.get("anchorOdo") is not None)
            else None)
    min_km = _num(opts.get("minKm"))
    if min_km is None:
        min_km = MIN_KM
    if dist is None or dist < min_km:
        return None
    used_pct = (open_t["anchorPct"] - open_t["lastPct"]
                if (not open_t.get("chargedSince")
                    and open_t.get("anchorPct") is not None
                    and open_t.get("lastPct") is not None)
                else None)
    kwh = (used_pct / 100) * cap if (used_pct is not None and used_pct > 0) else None
    mins = max(1, round((end_at - open_t["anchorAt"]) / 60000))
    mi = dist * MI_PER_KM
    anchor_pct = open_t.get("anchorPct")
    anchor_range_km = open_t.get("anchorRangeKm")
    anchor_temp_c = open_t.get("anchorTempC")
    return {
        "startedAt": open_t["anchorAt"],
        "endedAt": end_at,
        "minutes": mins,
        "distanceKm": _round(dist, 1),
        "distanceMi": _round(mi, 1),
        "usedPct": _round(used_pct, 1) if used_pct is not None else None,
        # raw start/end % (not just the usedPct delta) -- analytics.py's
        # range_accuracy() needs the actual starting % to compare against
        # what Kia displayed AT that %, not just how much was consumed.
        "startPct": _round(anchor_pct, 1) if anchor_pct is not None else None,
        "endPct": _round(open_t.get("lastPct"), 1) if open_t.get("lastPct") is not None else None,
        "kwh": _round(kwh, 2),
        "miPerKwh": _round(mi / kwh, 2) if (kwh is not None and kwh > 0) else None,
        "kwhPer100mi": _round(kwh / mi * 100, 1) if (kwh is not None and dist > 0) else None,
        "cost": _round(kwh * price, 2) if (kwh is not None and price > 0) else None,
        "pricePerKwh": price or None,
        "fromLat": open_t.get("anchorLat"), "fromLon": open_t.get("anchorLon"),
        "toLat": open_t.get("lastLat"), "toLon": open_t.get("lastLon"),
        "straightLineKm": _round(
            haversine_km(open_t.get("anchorLat"), open_t.get("anchorLon"),
                         open_t.get("lastLat"), open_t.get("lastLon")), 1),
        "chargedDuring": bool(open_t.get("chargedSince")),
        # Kia's own displayed EV range at the trip's START -- see
        # analytics.py's range_accuracy(). None if the car didn't report a
        # range reading at that moment.
        "startRangeKm": _round(anchor_range_km, 1) if anchor_range_km is not None else None,
        # outside temperature at the trip's start -- see analytics.py's
        # observed_efficiency() temperature buckets. The START reading, not
        # an average with the end: that's what the driver actually
        # experienced deciding whether to trust the range estimate.
        "outsideTempC": _round(anchor_temp_c, 1) if anchor_temp_c is not None else None,
    }


def update(open_t, cur, opts=None):
    """Fold one state sample into trip tracking.

    open_t: dict | None   trip in progress
    cur:    {"t", "odometerKm", "batteryPct", "charging", "carOn",
             "locationLat", "locationLon"}
    opts:   {"pricePerKwh", "capacityKwh", "minKm", "parkGapMin"}
    -> {"open": dict|None, "closed": dict|None}
    """
    opts = opts or {}
    t = _num(cur.get("t")) or (time.time() * 1000)
    odo = _num(cur.get("odometerKm"))
    pct = _num(cur.get("batteryPct"))
    lat = _num(cur.get("locationLat"))
    lon = _num(cur.get("locationLon"))
    range_km = _num(cur.get("rangeKm"))
    temp_c = _num(cur.get("outsideTempC"))
    charging = cur.get("charging") is True
    car_on = cur.get("carOn") is True
    gap_ms = (_num(opts.get("parkGapMin")) or PARK_GAP_MIN) * 60000

    if odo is None:
        return {"open": open_t, "closed": None}

    if not open_t:
        return {
            "open": {
                "anchorOdo": odo, "anchorPct": pct, "anchorAt": t,
                "anchorLat": lat, "anchorLon": lon,
                "anchorRangeKm": range_km, "anchorTempC": temp_c,
                "lastOdo": odo, "lastPct": pct, "lastAt": t,
                "lastLat": lat, "lastLon": lon,
                "lastRangeKm": range_km, "lastTempC": temp_c,
                "movedAt": t, "chargedSince": False,
            },
            "closed": None,
        }

    moved = odo > open_t["lastOdo"] + 0.05
    if charging or (pct is not None and open_t.get("lastPct") is not None
                    and pct > open_t["lastPct"] + 1):
        open_t["chargedSince"] = True

    if moved:
        open_t["lastOdo"] = odo
        open_t["movedAt"] = t
        if pct is not None:
            open_t["lastPct"] = pct
        if lat is not None:
            open_t["lastLat"] = lat
            open_t["lastLon"] = lon
        if range_km is not None:
            open_t["lastRangeKm"] = range_km
        if temp_c is not None:
            open_t["lastTempC"] = temp_c
        open_t["lastAt"] = t
        return {"open": open_t, "closed": None}

    if pct is not None:
        open_t["lastPct"] = pct
    if range_km is not None:
        open_t["lastRangeKm"] = range_km
    if temp_c is not None:
        open_t["lastTempC"] = temp_c
    if lat is not None and open_t["lastOdo"] == open_t["anchorOdo"]:
        open_t["anchorLat"] = lat
        open_t["anchorLon"] = lon
        if pct is not None:
            open_t["anchorPct"] = pct
        if range_km is not None:
            open_t["anchorRangeKm"] = range_km
        if temp_c is not None:
            open_t["anchorTempC"] = temp_c
    open_t["lastAt"] = t

    parked = (not car_on and t - open_t["movedAt"] >= gap_ms)
    if parked and open_t["lastOdo"] > open_t["anchorOdo"]:
        trip = _close(open_t, opts, open_t["movedAt"])
        return {
            "open": {
                "anchorOdo": open_t["lastOdo"], "anchorPct": open_t["lastPct"], "anchorAt": t,
                "anchorLat": open_t["lastLat"], "anchorLon": open_t["lastLon"],
                "anchorRangeKm": open_t.get("lastRangeKm"), "anchorTempC": open_t.get("lastTempC"),
                "lastOdo": open_t["lastOdo"], "lastPct": open_t["lastPct"], "lastAt": t,
                "lastLat": open_t["lastLat"], "lastLon": open_t["lastLon"],
                "lastRangeKm": open_t.get("lastRangeKm"), "lastTempC": open_t.get("lastTempC"),
                "movedAt": t, "chargedSince": False,
            },
            "closed": trip,
        }
    if parked:
        open_t["anchorOdo"] = open_t["lastOdo"]
        open_t["anchorPct"] = open_t["lastPct"]
        open_t["anchorAt"] = t
        open_t["anchorLat"] = open_t["lastLat"]
        open_t["anchorLon"] = open_t["lastLon"]
        open_t["anchorRangeKm"] = open_t.get("lastRangeKm")
        open_t["anchorTempC"] = open_t.get("lastTempC")
        open_t["chargedSince"] = False
    return {"open": open_t, "closed": None}


def summary(trips, days=30):
    cutoff = (time.time() * 1000) - days * 864e5
    km = kwh = cost = 0.0
    n = 0
    have_cost = have_kwh = False
    for tr in trips or []:
        if not tr or _num(tr.get("endedAt")) is None or tr["endedAt"] < cutoff:
            continue
        n += 1
        if tr.get("distanceKm") is not None:
            km += tr["distanceKm"]
        if tr.get("kwh") is not None:
            kwh += tr["kwh"]
            have_kwh = True
        if tr.get("cost") is not None:
            cost += tr["cost"]
            have_cost = True
    mi = km * MI_PER_KM
    return {
        "count": n,
        "distanceKm": _round(km, 1),
        "distanceMi": _round(mi, 1),
        "kwh": _round(kwh, 1) if have_kwh else None,
        "cost": _round(cost, 2) if have_cost else None,
        "miPerKwh": _round(mi / kwh, 2) if (have_kwh and kwh > 0) else None,
        "costPerMi": _round(cost / mi, 3) if (have_cost and mi > 0) else None,
    }
